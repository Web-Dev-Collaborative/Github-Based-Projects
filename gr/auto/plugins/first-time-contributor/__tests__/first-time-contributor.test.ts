import * as Auto from '@auto-it/core';
import makeCommitFromMsg from '@auto-it/core/dist/__tests__/make-commit-from-msg';
import Changelog from '@auto-it/core/dist/changelog';
import {
  makeChangelogHooks,
  makeHooks
} from '@auto-it/core/dist/utils/make-hooks';

import FirstTimeContributor from '../src';
import Octokit = require('@octokit/rest');

const graphql = jest.fn();
const exec = jest.fn();
exec.mockReturnValue('');
// @ts-ignore
jest.spyOn(Auto, 'execPromise').mockImplementation(exec);

const setup = (
  contributors: Partial<Octokit.ReposListContributorsResponseItem>[] = []
) => {
  const plugin = new FirstTimeContributor();
  const hooks = makeHooks();
  const changelogHooks = makeChangelogHooks();

  plugin.apply({
    hooks,
    git: {
      graphql,
      options: { repo: 'repo', owner: 'test' },
      github: {
        repos: {
          listContributors: () => Promise.resolve({ data: contributors })
        }
      }
    } as any
  } as Auto.Auto);
  hooks.onCreateChangelog.call(
    {
      hooks: changelogHooks,
      options: { baseUrl: 'https://github.com' }
    } as Changelog,
    Auto.SEMVER.patch
  );

  return changelogHooks;
};

describe('First Time Contributor Plugin', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should include all authors with no past contributors', async () => {
    const hooks = setup();
    const commits = [
      makeCommitFromMsg('foo', { name: 'Jeff', username: 'Jeff' }),
      makeCommitFromMsg('foo', { name: 'Andrew', username: 'Andrew' })
    ];

    graphql.mockReturnValueOnce({
      search: { issueCount: 0 }
    });
    graphql.mockReturnValueOnce({
      search: { issueCount: 0 }
    });

    expect(await hooks.addToBody.promise([], commits)).toMatchSnapshot();
  });

  test('should exclude a past contributor', async () => {
    const hooks = setup();
    const commits = [
      makeCommitFromMsg('foo', { name: 'Jeff', username: 'Jeff' }),
      makeCommitFromMsg('foo', { name: 'Andrew', username: 'Andrew' })
    ];

    graphql.mockReturnValueOnce({
      search: { issueCount: 1 }
    });

    expect(await hooks.addToBody.promise([], commits)).toMatchSnapshot();
  });

  test('should exclude a past contributor by username', async () => {
    const hooks = setup([{ login: 'Andrew420' }]);
    const commits = [
      makeCommitFromMsg('foo', { name: 'Jeff123', username: 'Jeff123' }),
      makeCommitFromMsg('foo', { name: 'Andrew420', username: 'Andrew420' })
    ];

    graphql.mockReturnValueOnce({
      search: { issueCount: 1 }
    });

    expect(await hooks.addToBody.promise([], commits)).toMatchSnapshot();
  });

  test('should not include past contributors', async () => {
    const hooks = setup();
    const commits = [
      makeCommitFromMsg('foo', { name: 'Jeff', username: 'Jeff' }),
      makeCommitFromMsg('foo', { name: 'Andrew', username: 'Andrew' })
    ];

    graphql.mockReturnValue({
      search: { issueCount: 2 }
    });

    expect(await hooks.addToBody.promise([], commits)).toMatchSnapshot();
  });

  test('should include a username if it exists', async () => {
    const hooks = setup();
    const commits = [
      makeCommitFromMsg('foo', { name: 'Jeff', username: 'jeff-the-snake' }),
      makeCommitFromMsg('foo', { name: 'Andrew', username: 'hipstersmoothie' })
    ];

    graphql.mockReturnValue({
      search: { issueCount: 1 }
    });

    expect(await hooks.addToBody.promise([], commits)).toMatchSnapshot();
  });

  test('should include only username if no name exists', async () => {
    const hooks = setup();
    const commits = [
      makeCommitFromMsg('foo', { username: 'jeff-the-snake', name: '' })
    ];

    graphql.mockReturnValue({
      search: { issueCount: 1 }
    });

    expect(await hooks.addToBody.promise([], commits)).toMatchSnapshot();
  });
});
