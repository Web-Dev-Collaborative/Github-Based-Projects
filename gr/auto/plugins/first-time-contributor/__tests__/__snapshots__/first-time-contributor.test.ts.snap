// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First Time Contributor Plugin should exclude a past contributor 1`] = `
Array [
  ":tada: This release contains work from a new contributor! :tada:

Thank you, Jeff ([@Jeff](https://github.com/Jeff)), for all your work!",
]
`;

exports[`First Time Contributor Plugin should exclude a past contributor by username 1`] = `
Array [
  ":tada: This release contains work from a new contributor! :tada:

Thank you, Jeff123 ([@Jeff123](https://github.com/Jeff123)), for all your work!",
]
`;

exports[`First Time Contributor Plugin should include a username if it exists 1`] = `
Array [
  ":tada: This release contains work from new contributors! :tada:

Thanks for all your work!

:heart: Jeff ([@jeff-the-snake](https://github.com/jeff-the-snake))

:heart: Andrew ([@hipstersmoothie](https://github.com/hipstersmoothie))",
]
`;

exports[`First Time Contributor Plugin should include all authors with no past contributors 1`] = `
Array [
  ":tada: This release contains work from new contributors! :tada:

Thanks for all your work!

:heart: Jeff ([@Jeff](https://github.com/Jeff))

:heart: Andrew ([@Andrew](https://github.com/Andrew))",
]
`;

exports[`First Time Contributor Plugin should include only username if no name exists 1`] = `
Array [
  ":tada: This release contains work from a new contributor! :tada:

Thank you, [@jeff-the-snake](https://github.com/jeff-the-snake), for all your work!",
]
`;

exports[`First Time Contributor Plugin should not include past contributors 1`] = `Array []`;
